### ISSUE TITLE
MTKv10 discussion

### ISSUE THREAD

**Issue by AayushSabharwal** [SciML/ModelingToolkit.jl#3204]
This issue is meant to track and discuss points of action and potential breaking changes for a future MTKv10 release

- Make `@named` _always_ wrap symbolic arguments in `ParentScope`, so they can be passed multiple levels down the hierarchy. This will need more consideration after identifying which tests break upon making this change.
- Discuss `complete` semantics
- Enforce the `u0map` argument only contains variables, and `pmap` only parameters. Discretes will need an exception here.

**Comment by AayushSabharwal** [comment]
- Remove the find/replace stuff in `expand_connections`. This is only used for MTKStdlib's implementation of `AnalysisPoint` but that is moved to MTK as of #3285.

**Comment by baggepinnen** [comment]
I'd like to discuss making the syntax
```
Component(inner_par = outer_par)
```
Introduce a parameter dependence instead of a "dependent default". I find the current behavior is almost never what I want. In modelica, this syntax introduces the equivalent of a parameter dependence.

**Comment by vyudu** [comment]
Possible changes to field names in symbolic metadata:
- `vartype` becoming `variable_source` 
- new field `variable_type` (corresponding to the MTK variable type)

**Comment by AayushSabharwal** [comment]
^ This specifically refers to the `NamedTuple` returned from `get_variable_metadata`

**Comment by AayushSabharwal** [comment]
Other discussions worth having:
- defaults are now just soft constraints that are only respected sometimes. Especially for parameters since they have to opt-in to initialization. 
- Not relying on variable metadata as much. This refers to multiple things
  - `@mtkmodel` macro using it for keeping track of defaults
  - More importantly, `structural_simplify` using it to figure out what is a parameter and what isn't. `ODESystem` can use the metadata to do this, but `structural_simplify` should respect whatever bucket we put it in.

**Comment by hersle** [comment]
> Discuss `complete` semantics

Should `complete()` be changed to use `flatten = false` by default? I find it undesireable that the current default `flatten = true` destroys the hierarchical model structure without me explicitly requesting it. There are some bugs with `flatten = false`, though (https://github.com/SciML/ModelingToolkit.jl/issues/3322).

**Comment by hersle** [comment]
> More importantly, `structural_simplify` using it to figure out what is a parameter and what isn't. ODESystem can use the metadata to do this, but `structural_simplify` should respect whatever bucket we put it in.

I always thought it could be simpler to have just the `@variables` macro, and that it should be sufficient to write
```julia
@variables t # instead of @independent_variables
vars = @variables x(t)
pars = @variables P # instead of @parameters
eqs = [...]
@named sys = ODESystem(eqs, t, vars, pars)
```
because it should be clear from the arguments to `ODESystem` what the independent variable, unknowns and parameters are. Maybe this breaks stuff with the "convenience dispatch" `ODESystem(eqs, t)` and the `@mtkmodel` macro (which I have less experience with), or causes other problems, though. This is not a huge deal in practice, though, and I apologize if this is very naive suggestion 😅

**Comment by AayushSabharwal** [comment]
> Maybe this breaks stuff with the "convenience dispatch" ODESystem(eqs, t)

It doesn't break stuff, but it makes catching user errors more difficult. Maybe they accidentally create `@variables x` intending it to be an unknown?

> the `@mtkmodel` macro 

`@mtkmodel` does its own parsing (something I strongly dislike) so that wouldn't be an issue

> or causes other problems, though.

It causes problems for `NonlinearSystem`, for example, where `@variables x` could just as well be a parameter as an unknown without the metadata.

**Comment by AayushSabharwal** [comment]
https://github.com/SciML/ModelingToolkit.jl/pull/3335#issuecomment-2597825918

**Comment by ChrisRackauckas** [comment]
> defaults are now just soft constraints that are only respected sometimes.

But they are always enforced at the start of a solve?

> More importantly, structural_simplify using it to figure out what is a parameter and what isn't. ODESystem can use the metadata to do this, but structural_simplify should respect whatever bucket we put it in.

yes, because otherwise it couldn't handle `D(x) ~ 0` turning into a parameter.

**Comment by AayushSabharwal** [comment]
https://github.com/SciML/ModelingToolkit.jl/pull/3354

**Comment by vyudu** [comment]
Requiring `Pre` in callback designations

**Comment by SebastianM-C** [comment]
https://github.com/SciML/ModelingToolkit.jl/issues/3444#issuecomment-2712081409

Would it make sense to have `Initial`s and other things that might be introduced by MTK in `full_parameters` and keep `parameters` for the user specified ones?

**Comment by ChrisRackauckas** [comment]
Not really, because "user specified ones" is somewhat undefined anyways. If you do `D(x) ~ 0`, then structural simplify should eliminate `x` as a part of the state and instead make it a parameter and not solve it, right? While that's not done right now, that's one that clearly should be shown. That's only a start too, if we start to integrate the analytical solvers which we are planning to do, other promotions to parameters would be there as well. In which case, it's not clear the user would just want the "user specified ones", but instead the things that are captured as parameters.

**Comment by SebastianM-C** [comment]
Sure, but in that case can't we export full_parameters and add more options to that? It seems intuitive (to me) that parameters would be the most narrow set and then full_parameters has several kwargs that allow you to see more of the promoted or automatically added ones.

**Comment by baggepinnen** [comment]
> If you do D(x) ~ 0, then structural simplify should eliminate x as a part of the state and instead make it a parameter and not solve it, right?

we rely on `x` being kept in the state in this case in a lot of places. We might change `x` externally using the integrator interface, or generate a control function for a state estimator that estimates `x` from data etc.

**Comment by ChrisRackauckas** [comment]
`x` changing in a callback is just a discrete-time state value but is kept internally in MTK as a parameter of sorts. We can know from symbolic callbacks whether it should be treated as discrete-time or constant, and the user can set it as irreducible if it's doing something odd only in function callbacks.

**Comment by AayushSabharwal** [comment]
https://github.com/SciML/ModelingToolkit.jl/issues/3539

**Comment by baggepinnen** [comment]
Should we take the opportunity to change names that have been poorly chosen? Like
- `structural_simplify` -> `simplify_structure` or `structural_simplification` so it sounds like English, or `mtkbuild` to align with `@mtkbuild`
- `generate_xxx / build_xxx` -> a consistent choice of one of them

**Comment by AayushSabharwal** [comment]
I like `mtkbuild`

**Comment by ChrisRackauckas** [comment]
Generate functions and build functions are different and consistent.

**Comment by AayushSabharwal** [comment]
`generate` did a somewhat weird thing of returning an RGF compiled in Symbolics if you pass `expression = Val{false}`. This doesn't use our `eval_expression` and `expression_module` stuff, and requires splatting `MTKParameters` to use. In #3543 I've made it so that passing `expression = Val{false}` to `generate_*` functions returns a compiled function respecting `eval_expression` etc. and callable without splatting. This also improves `*Function` constructor code.

**Comment by ChrisRackauckas** [comment]
We could collapse build and generate functions if we just make that difference based on `expression`. That's a reasonable unification with other parts of the interface.

**Comment by AayushSabharwal** [comment]
Expose `tearing_substitute_expr` under a better name - it basically substitutes observed equations into an expression and is a useful utility.

**Comment by AayushSabharwal** [comment]
- Remove `check_length = false` as a keyword argument to `XProblem`

**Comment by asinghvi17** [comment]
Should this be closed?
